framework/src/userobject/ElementSubdomainModifier.C

```diff
@@ -0,0 +1,347 @@
+  mesh.getMesh().get_boundary_info().sideset_name(_moving_boundary_id) = _moving_boundary_name;
+  mesh.getMesh().get_boundary_info().nodeset_name(_moving_boundary_id) = _moving_boundary_name;
+}
+
+void
+ElementSubdomainModifier::initialize()
+{
+  _moved_elems.clear();
+  _moved_displaced_elems.clear();
+  _moved_nodes.clear();
+}
+
+void
+ElementSubdomainModifier::execute()
+{
+  // First, compute the desired subdomain ID for the current element.
+  SubdomainID subdomain_id = computeSubdomainID();
+
+  // If the current element's subdomain ID isn't what we want
+  if (subdomain_id != std::numeric_limits<SubdomainID>::max() &&
+      _current_elem->subdomain_id() != subdomain_id)
+  {
+    // Current element ID, used to index both the element on the displaced and undisplaced meshes.
+    dof_id_type elem_id = _current_elem->id();
+    // Change the element's subdomain
+    Elem * elem = _mesh.elemPtr(elem_id);
+    Elem * displaced_elem =
+        _displaced_problem ? _displaced_problem->mesh().elemPtr(elem_id) : nullptr;
+
+    // Save the affected nodes so that we can later update/initialize the solution
+    for (unsigned int i = 0; i < elem->n_nodes(); ++i)
+      _moved_nodes.insert(elem->node_id(i));
+
+    elem->subdomain_id() = subdomain_id;
+    _moved_elems.push_back(elem);
+    if (displaced_elem)
+    {
+      displaced_elem->subdomain_id() = subdomain_id;
+      _moved_displaced_elems.push_back(displaced_elem);
+    }
+  }
+}
+
+void
+ElementSubdomainModifier::finalize()
+{
+  /*
+    Synchronize ghost element subdomain ID
+    Note: this needs to be done before updating boundary info because
+    updating boundary requires the updated element subdomain ids
+  */
+  SyncSubdomainIds sync(_mesh.getMesh());
+  Parallel::sync_dofobject_data_by_id(_mesh.getMesh().comm(),
+                                      _mesh.getMesh().elements_begin(),
+                                      _mesh.getMesh().elements_end(),
+                                      sync);
+  updateBoundaryInfo(_mesh, _moved_elems);
+
+  // Similarly for the displaced mesh
+  if (_displaced_problem)
+  {
+    SyncSubdomainIds sync_displaced(_displaced_problem->mesh().getMesh());
+    Parallel::sync_dofobject_data_by_id(_displaced_problem->mesh().getMesh().comm(),
+                                        _displaced_problem->mesh().getMesh().elements_begin(),
+                                        _displaced_problem->mesh().getMesh().elements_end(),
+                                        sync_displaced);
+    updateBoundaryInfo(_displaced_problem->mesh(), _moved_displaced_elems);
+  }
+
+  // Reinit equation systems
+  _fe_problem.meshChanged();
+
+  if (_apply_ic)
+  {
+    // Apply initial condition for the newly moved elements and boundary nodes
+    buildMovedElemsRange();
+    buildMovedBndNodesRange();
+    _fe_problem.projectInitialConditionOnCustomRange(movedElemsRange(), movedBndNodesRange());
```

Comment:
I don't know if I need to also project initial conditions on displaced elements. I am only projecting ICs on undisplaced elements.

Reply:
I haven't seen any issues so far without initializing solutions on the displaced mesh. But as you said there maybe corner cases I'm not aware of. I'll just rely on `syncSolutions()` for now.

Reply:
Hold on, I actually got segfault if I call `_displaced_problem->syncSolutions()`. Not sure why.